toolchain/check/handle_binding_pattern.cpp

```diff
@@ -247,6 +247,18 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
           }
           break;
         }
+        case Lex::TokenKind::Choice:
+          // Check if we're building a pattern for a choice alternative, not the
+          // choice type itself.
+          if (context.node_stack()
+                  .PeekContains<Parse::NodeKind::ChoiceDefinitionStart>()) {
+            // Implicit param lists are prevented during parse.
+            CARBON_CHECK(
+                context_node_kind != Parse::NodeKind::ImplicitParamListStart,
+                "choice alternative with implicit parameters");
+            break;
+          }
+          [[fallthrough]];
```

Comment:
Might it make sense to drop this code (as well as `NodeStack::PeekContains`)?

```suggestion
        case Lex::TokenKind::Choice:
```

If I understand correctly, all we'd lose is the `CHECK`, and it's not clear to me that that invariant is brittle enough (or dangerous enough if broken) to be worth going to this much trouble.

Reply:
I used to have a diagnostic here, but as you can see it wasn't needed. But this is also providing a `break` in the alternatives case. It will probably need to check other things more like Fn once we have parameters. But self may be allowed.

I can do the check here though using scope_stack() instead of PeekContains(), so I've removed PeekContains(). And removed the CHECK.

Reply:
Oh, sorry, I missed the `break`. That being the case, feel free to put the `CHECK` back.

Reply:
Done.